### PR TITLE
Add `audio/general/text_to_speech` project setting to enable/disable TTS.

### LINKED ISSUE #77124: Text to speech draining CPU with Pulse Audio when not in use
### Godot version

_Present in 3.x and master, I've milestoned for 3.6 as I'm investigating in 3.x but same will apply in master._
3.x current (16 may 2023) (567128b), 4.x (last week)

### System information

Linux Mint 20.1

### Issue description

_Note - This issue is related (but not the same) as #55608 . That issue was due to the engine using Pulse Audio, this is due to text to speech (TTS)._

System monitor shows that running even the project manager, speech dispatcher is using Pulse Audio which uses significant CPU (11% in my case) even when Godot audio output is switched off (using dummy driver).

By using `pactl list clients` you can list which apps are causing PulseAudio to use CPU.

This info here regards 3.x, but reports indicate the same problem in 4.x. See https://github.com/godotengine/godot/issues/55608#issuecomment-1540384278 .

The normal for the engine audio output is:
```
Client #51
	Driver: protocol-native.c
	Owner Module: 12
	Properties:
		application.name = "Godot Engine Editor"
		native-protocol.peer = "UNIX socket client"
		native-protocol.version = "33"
		application.process.id = "9158"
		application.process.user = "baron"
		application.process.host = "baron-fanless"
		application.process.binary = "Godot_v3.6-beta1_x11.64"
		application.language = "C"
		window.x11.display = ":0"
		application.process.machine_id = ""
		application.process.session_id = "c2"
```
When text to speech is active you get this:
```
Client #54
	Driver: protocol-native.c
	Owner Module: 12
	Properties:
		application.name = "speech-dispatcher-espeak-ng"
		native-protocol.peer = "UNIX socket client"
		native-protocol.version = "33"
		application.process.id = "9257"
		application.process.user = "baron"
		application.process.host = "baron-fanless"
		application.process.binary = "sd_espeak-ng"
		application.language = "C"
		window.x11.display = ":0"
		application.process.machine_id = ""
		application.process.session_id = "c2"

Client #55
	Driver: protocol-native.c
	Owner Module: 12
	Properties:
		application.name = "speech-dispatcher-generic"
		native-protocol.peer = "UNIX socket client"
		native-protocol.version = "33"
		application.process.id = "9267"
		application.process.user = "baron"
		application.process.host = "baron-fanless"
		application.process.binary = "sd_generic"
		application.language = "C"
		window.x11.display = ":0"
		application.process.machine_id = ""
		application.process.session_id = "c2"

Client #56
	Driver: protocol-native.c
	Owner Module: 12
	Properties:
		application.name = "speech-dispatcher-dummy"
		native-protocol.peer = "UNIX socket client"
		native-protocol.version = "33"
		application.process.id = "9270"
		application.process.user = "baron"
		application.process.host = "baron-fanless"
		application.process.binary = "sd_dummy"
		application.language = "C"
		window.x11.display = ":0"
		application.process.machine_id = ""
		application.process.session_id = "c2"
```

This problem may also occur in different form on other platforms, but both my and @wb14123 's reports are on Linux, so can only confirm there so far.

### Steps to reproduce

Just run the engine to get the project manager, examine system monitor to see CPU use per process, and run `pactl list clients`.

What makes this more tricky is that this problem appears to be intermittent. I haven't quite pinned down why it is occurring some runs / builds and not others, will add more info as it becomes apparent.

EDIT: Ah the intermittent nature may be because I get `Warning: Speech Dispatcher development libraries not found. Disabling Text-to-Speech support.` on local builds, but production builds presumably are compiling with text to speech, and getting the problem.

### Minimal reproduction project

N/A

### Discussion
The issue appears to be caused because Godot's TTS is always on. It will probably be fixed if we can add a project setting whether to use TTS or not. See: 
https://github.com/godotengine/godot/issues/55608#issuecomment-1545194612

Presumably the issue is caused by #61316 / #56192 .

### PR DESCRIPTION
**Pull request by bruvzg**
Adds setting to completely disable TTS support.

The issue seems to be Linux specific, but for consistency, the setting is applied on all systems.

Fixes https://github.com/godotengine/godot/issues/77124

### CODE REVIEW

File: core/config/project_settings.cpp [hunk 1]
```diff
@@ -1258,6 +1258,8 @@ ProjectSettings::ProjectSettings() {
 	GLOBAL_DEF("application/config/custom_user_dir_name", "");
 	GLOBAL_DEF("application/config/project_settings_override", "");
 
+	GLOBAL_DEF_RST("application/config/text_to_speech", true);
```
Comment (akien-mga) [inline core/config/project_settings.cpp:1261]: I wonder if `application/config/` is the best location for this. Currently it's used for very general project config (name, main scene, user dir, etc.).

Might be better under `audio/driver/` or `audio/general/`?

**Comment by lawnjelly** [comment]
On principle the default should be false imo.

Users will not expect text to speech to be using resources and CPU when they have never requested it - slowing down their game or eating battery unnecessarily. We don't for example, render shadow maps just in case user might want shadows.

If this is to be used for editor accessibility (I don't know if it is currently), we could consider a separate editor setting. Also useful would be a runtime command line argument to override these settings (in case end user wants to disable text to speech).

Another approach once the project setting is working, as @bruvzg suggested previously, is simply allow deferring speech startup until a speech command is issued, and then the default could perhaps be changed to on, if there was no cost. But in any case having it switch-offable is eminently desirable.

**Comment by bruvzg** [comment]
> On principle the default should be false imo.

I would set it to `false`, but this will break existing projects with TTS (but I do not think there are many, so might be ok for 4.1).

> Users will not expect text to speech to be using resources and CPU when they have never requested it - slowing down their game or eating battery unnecessarily. We don't for example, render shadow maps just in case user might want shadows.

It seems to Linux specific, and might be specific to speech dispatcher or Pulse Audio version (I do not get more the 1.5% of usage).

> If this is to be used for editor accessibility (I don't know if it is currently), we could consider a separate editor setting. 

It's not used for anything in the editor, and not a good way to implement accessibility (I'm working on a proper screen reader support right now - https://github.com/godotengine/godot/pull/76829).

> is simply allow deferring speech startup until a speech command is issued

Not sure if it's worth overcomplicating it.

File: core/config/project_settings.cpp (hunk 1 above)
Comment (bruvzg) [reply core/config/project_settings.cpp]: Moved to `audio/general/text_to_speech`.

**Comment by lawnjelly** [comment]
> I would set it to false, but this will break existing projects with TTS (but I do not think there are many, so might be ok for 4.1).

Surely all it will take is those few users flipping a project setting when they next export the project? Provided this is included in the release notes, this seems unlikely to cause many problems (it should be fairly obvious if a feature they are using is no longer working).

There's no change to e.g. code needed. This seems imo by far the lesser of two evils, if the alternative is making everyone else pay for this oversight, in perpetuity.

**Comment by bruvzg** [comment]
Well, error messages directly say what to do, so I guess it's OK. Changed default to `false`.

**Review by lawnjelly — APPROVED** [review]
Looks ok to me, I'm not an expert on all the platforms so any second pair of eyes to check would be good. :+1: 

Also reminder we need an 3.x version too, I guess it should mostly be the same.

**Comment by akien-mga** [comment]
Looks good to me. The commit message should be amended like I just did with the PR title.

Also, I think @lawnjelly's concern also includes the CPU usage of the editor, which this wouldn't address as it's only for the project settings so project runtime. So this might also need an editor setting to disable editor TTS.
It could be disabled by default until we add accessibility features using it (in which case it should be enabled by default IMO).

File: platform/android/display_server_android.cpp [hunk 2]
```diff
@@ -531,6 +531,11 @@ void DisplayServerAndroid::notify_surface_changed(int p_width, int p_height) {
 DisplayServerAndroid::DisplayServerAndroid(const String &p_rendering_driver, DisplayServer::WindowMode p_mode, DisplayServer::VSyncMode p_vsync_mode, uint32_t p_flags, const Vector2i *p_position, const Vector2i &p_resolution, int p_screen, Error &r_error) {
 	rendering_driver = p_rendering_driver;
 
+	bool tts_enabled = GLOBAL_GET("audio/general/text_to_speech");
+	if (tts_enabled) {
+		TTS_Android::init();
+	}
+
```
Comment (m4gr3d) [inline platform/android/display_server_android.cpp:538]: Shouldn't this be in `tts_android` or in the java layer itself?
While this is not the case now, I'm primarily concerned about any changes that may alter the init order of `TTS_Android` and `DisplayServerAndroid` and thus create a bug.

**Review by m4gr3d — CHANGES_REQUESTED** [review]

File: platform/android/display_server_android.cpp (hunk 2 above)
Comment (bruvzg) [reply platform/android/display_server_android.cpp:538]: `TTS_Android` is directly used only from display server, the rest of code calls it via display server wrappers. And `GodotTTS` on the Java side is created before project setting are available.

Comment (m4gr3d) [reply platform/android/display_server_android.cpp:538]: `TTS_Android` is initialized in [`java_godot_lib_jni.cpp`](https://github.com/godotengine/godot/blob/437041a983a81f9768673dc5374f644634f940f9/platform/android/java_godot_lib_jni.cpp#L136) via [`TTS_Android::setup(...)`](https://github.com/godotengine/godot/blob/437041a983a81f9768673dc5374f644634f940f9/platform/android/tts_android.cpp#L51).

The check for `tts_enabled` and the invocation of `TTS_Android::init()` should be done within that method, after the jni methods have been set up, so we can consolidate all the init logic in one location.

Comment (bruvzg) [reply platform/android/display_server_android.cpp:538]: > The check for tts_enabled and the invocation of TTS_Android::init() should be done within that method, after the jni methods have been set up, so we can consolidate all the init logic in one location.

That won't work, project settings are loaded in `Main::setup`, so it's not loaded in the `GodotLib.initialize`, but  it should be possible to move it to the `GodotLib.setup`.

Comment (bruvzg) [reply platform/android/display_server_android.cpp]: Moved init to `GodotLib.setup` and `TTS_Android::setup`.

**Review by m4gr3d — APPROVED** [review]
The Android section looks good, thanks for the update!

**Comment by YuriSizov** [comment]
@bruvzg Please amend the commit message, it still incorrectly refers to the setting name. Should be `audio/general/text_to_speech`.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Would need a `3.x` backport if we want to make it opt-in for 3.6 too.

**Pull request merged**